### PR TITLE
CurrentAddress - mark first address used and return new address

### PR DESCRIPTION
**Pull request by tuxcanfly**
Fixed `Last(In/Ex)ternalAddress` to make sure `CurrentAddress` returns a new address when the first and subsequent addresses are flagged as used.

### CODE REVIEW

**Comment by manan19** [comment]
This solves part of the problem I was facing.

I think the first generated external address should be also be added to first rescan job. 

Right now, despite calling .CurrentAddress, the first address isn't included in the rescan job.
Since it's not included in that list, it won't be marked as used, even though it has a balance on it. .CurrentAddress will keep returning the same address.

This issue might need to be fixed in the wallet package.

Also, IMO .NewAddress should be changed to .newAddress (un-exported)

**Comment by jrick** [comment]
What about erroring on last in/external address if none have been created yet?  It should never have been creating new addresses to begin with.

**Comment by manan19** [comment]
@jrick That will work too. It means that .CurrentAddress can implement the error handling for .LastExternalAddress and call .NewAddress.

**Comment by jrick** [comment]
The rescan with 0 addresses is actually intentional because the response that btcwallet gets after the finished rescan is the block that the wallet is marked in sync with.

However, btcd should detect the 0 address 0 outpoint rescan and not actually load blocks and compare addresses against empty maps in that case.

File: waddrmgr/manager.go [hunk 1]
```diff
@@ -1580,7 +1580,11 @@ func (m *Manager) LastExternalAddress(account uint32) (ManagedAddress, error) {
 		return nil, err
 	}
 
-	return acctInfo.lastExternalAddr, nil
+	if acctInfo.nextExternalIndex > 0 {
+		return acctInfo.lastExternalAddr, nil
+	}
+
+	return nil, managerError(ErrAddressNotFound, "address not found", nil)
```
Comment (jrick) [inline waddrmgr/manager.go]: Since this isn't looking up a particular address by string, hash160, etc., can you change the error description string to something like "no previous external address"?

Also, need a similar error for LastInternalAddress below.

**Comment by jrick** [comment]
Please update the PR title and description, since this problem is being fixed by a different method.

**Comment by tuxcanfly** [comment]
Yup, I agree about not creating addresses before they are required.

Updated `Last(In/Ex)ternalAddress` to err on first  address. Also had to update `loadAccountInfo` to query whether the addr is used and clear the cache after it's flagged.

File: waddrmgr/db.go [hunk 2]
```diff
@@ -1031,6 +1019,18 @@ func fetchAddressByHash(tx walletdb.Tx, addrHash []byte) (interface{}, error) {
 	return nil, managerError(ErrDatabase, str, nil)
 }
 
+// fetchAddressUsed returns true if the provided address id was flagged as used.
+func fetchAddressUsed(tx walletdb.Tx, addressID []byte) bool {
```
Comment (jrick) [inline waddrmgr/db.go]: As discussed on IRC, I'd like to see all usage of the phrase "address ID" replaced with something like "Hash160", since it is the ripemd160 after sha256 of the pubkey, or the script.  This is ok for now, but please fix this in another pr.

File: waddrmgr/db.go [hunk 3]
```diff
@@ -1031,6 +1019,18 @@ func fetchAddressByHash(tx walletdb.Tx, addrHash []byte) (interface{}, error) {
...
+	bucket := tx.RootBucket().Bucket(usedAddrBucketName)
+
+	addrHash := fastsha256.Sum256(addressID)
+	val := bucket.Get(addrHash[:])
+	if val != nil {
```
Comment (jrick) [inline waddrmgr/db.go]:
```Go
return bucket.Get(addrHash[:]) != nil
```

File: waddrmgr/manager.go [hunk 4]
```diff
@@ -1360,6 +1360,16 @@ func (m *Manager) Unlock(passphrase []byte) error {
 	return nil
 }
 
+// fetchUsed returns true if the provided address id was flagged used.
+func (m *Manager) fetchUsed(addressID []byte) bool {
+	var used bool
+	_ = m.namespace.View(func(tx walletdb.Tx) error {
```
Comment (jrick) [inline waddrmgr/manager.go]: Shouldn't ignore the error here.

File: waddrmgr/manager.go [hunk 5]
```diff
@@ -686,12 +686,12 @@ func (m *Manager) Address(address btcutil.Address) (ManagedAddress, error) {
 	return m.loadAndCacheAddress(address)
 }
 
-// AddrAccount returns the account to which the given address belongs.
-func (m *Manager) AddrAccount(address btcutil.Address) (uint32, error) {
+// AddrAccount returns the account to which the given address id belongs.
+func (m *Manager) AddrAccount(addressID []byte) (uint32, error) {
```
Comment (jrick) [inline waddrmgr/manager.go]: Why was this changed from taking the address to the hash160?

**Comment by manan19** [comment]
The way .CurrentAddress is implemented it would generate only 1 .NewAddress and stop. I don't think it will work well, when a wallet is being restored from a seed.

1st address, 0 balance, 2 transactions
change address, 0.1 balance, 1 transaction
2nd ext address, 0.2 balance, 1 transaction

.CurrentAddress in this scenario would be expected to return the 3rd ext address. It should recursively generate a new address until it comes across an address with 0 balance and 0 transactions. 

(I've assumed change addresses are generated from a different branch and hence the numbers)

**Comment by jrick** [comment]
CurrentAddress only exists to implement the currentaddress rpc.  It's ok if it's not terribly useful for recovering from a seed, since it wasn't designed for that use case.

File: waddrmgr/db.go (hunk 2 above)
Comment (tuxcanfly) [inline waddrmgr/db.go]: OK. I'm going to keep any new vars in this PR as addressID so I can find and replace all of them in the other PR.

File: waddrmgr/manager.go (hunk 4 above)
Comment (tuxcanfly) [inline waddrmgr/manager.go]: Ok, the only err possible here is `ErrBucketNotFound` which shouldn't happen (since the bucket is created at setup) but some other walletdb backend might return another err.

Comment (jrick) [inline waddrmgr/manager.go]: Or the db is unreachable, or mmap fails, or.... :)

File: waddrmgr/manager.go [hunk 6]
```diff
@@ -1360,15 +1363,31 @@ func (m *Manager) Unlock(passphrase []byte) error {
...
+	err := m.namespace.View(func(tx walletdb.Tx) error {
+		used = fetchAddressUsed(tx, addressID)
+		return nil
+	})
+	return used, err
+}
+
+// MarkUsed updates the used flag for the provided address.
+func (m *Manager) MarkUsed(address btcutil.Address) error {
+	addressID := address.ScriptAddress()
```
Comment (tuxcanfly) [inline waddrmgr/manager.go]: @jrick I think this was the external API call you mentioned which takes `addressID`

File: waddrmgr/manager.go (hunk 4 above)
Comment (tuxcanfly) [inline waddrmgr/manager.go]: I see, will keep that in mind.

File: waddrmgr/manager.go (hunk 6 above)
Comment (jrick) [inline waddrmgr/manager.go]: as long as it's a different type, finding and unbreaking callers is easy.

File: waddrmgr/manager.go [hunk 7]
```diff
@@ -1360,15 +1363,31 @@ func (m *Manager) Unlock(passphrase []byte) error {
...
 	if err != nil {
 		return maybeConvertDbError(err)
 	}
-	// 'used' flag has become stale so remove key from cache
+	// Clear caches which might have stale entries for used addresses
+	account, err := m.AddrAccount(address)
+	if err != nil {
+		return err
+	}
+	delete(m.acctInfo, account)
```
Comment (tuxcanfly) [inline waddrmgr/manager.go]: @jrick This is a bit of a hack since `acctInfo` caches the derived last (ex/in) addresses. I guess a better solution would be to stop deriving addrs and always load them from the db.

Comment (davecgh) [inline waddrmgr/manager.go]: I'd advise against this.  It's actually faster to rederive them than it is to load and deserialize them.

**Comment by tuxcanfly** [comment]
@jrick @davecgh I think the larger problem is that we're storing the mutable `used` in an otherwise immutable address and this leads to issues like the above where we need to be careful about clearing the caches. I just started looking at wtxmgr but if we could index transactions by address, it might be easier to check that we have no transactions belonging to the address and hence determine it's used or not.

File: waddrmgr/manager.go (hunk 7 above)
Comment (tuxcanfly) [inline waddrmgr/manager.go]: OK, I'll leave it to derive the addrs for now.

**Comment by jrick** [comment]
I think the used flag is still correct.  If a transaction is sent to some address, and the address is marked "used", then if that transaction is removed from wtxmgr (perhaps it was double spent), the address shouldn't become unused.  Once used, the address should never become unused again.

**Comment by tuxcanfly** [comment]
Yeah, just checked wtxmgr and looks like it is ranging over all transactions.

OK, let me test this out and prep for review. Thanks.

**Comment by jrick** [comment]
Where is it ranging transactions?  Only place I can remember it doing that and filtering by address is `Wallet.ListAddressTransactions`, which I want to remove (along with the RPC).  See #231

**Comment by jrick** [comment]
Well, let's keep things simple.  Until waddrmgr and wtxmgr are properly combined (this will happen after the current wtxmgr integration is merged), we'll keep the transaction iterating to detect used-ness.  Once combined, we'll introduce this flag again.

**Comment by tuxcanfly** [comment]
Yeah, I was checking if `ListAddressTransactions`  was filtering by address or if it was using an index.

File: waddrmgr/manager_test.go [hunk 8]
```diff
@@ -1028,28 +1028,32 @@ func testMarkUsed(tc *testContext) bool {
 	addrHash := expectedAddr1.addressHash
 	addr, err := btcutil.NewAddressPubKeyHash(addrHash, chainParams)
 
+	maddr, err := tc.manager.Address(addr)
+	if err != nil {
+		tc.t.Errorf("%s: unexpected error: %v", prefix, err)
```
Comment (jrick) [inline waddrmgr/manager_test.go]: Can't continue with the rest of this test, either return here of make this fatal

File: waddrmgr/manager_test.go [hunk 9]
```diff
@@ -1028,28 +1028,32 @@ func testMarkUsed(tc *testContext) bool {
...
+	}
 	if tc.create {
 		// Test that initially the address is not flagged as used
-		maddr, err := tc.manager.Address(addr)
+		used, err := maddr.Used()
 		if err != nil {
 			tc.t.Errorf("%s: unexpected error: %v", prefix, err)
```
Comment (jrick) [inline waddrmgr/manager_test.go]: return or fatal

File: waddrmgr/manager_test.go [hunk 10]
```diff
@@ -1028,28 +1028,32 @@ func testMarkUsed(tc *testContext) bool {
...
+		if used != false {
 			tc.t.Errorf("%v: unexpected used flag -- got "+
-				"%v, want %v", prefix, maddr.Used(), expectedAddr1.used)
+				"%v, want %v", prefix, used, expectedAddr1.used)
 		}
 	}
-	err = tc.manager.MarkUsed(addrHash)
+	err = tc.manager.MarkUsed(addr)
 	if err != nil {
 		tc.t.Errorf("%s: unexpected error: %v", prefix, err)
```
Comment (jrick) [inline waddrmgr/manager_test.go]: return or fatal

File: waddrmgr/manager_test.go [hunk 11]
```diff
@@ -1028,28 +1028,32 @@ func testMarkUsed(tc *testContext) bool {
...
 	}
-	maddr, err := tc.manager.Address(addr)
+	used, err := maddr.Used()
 	if err != nil {
 		tc.t.Errorf("%s: unexpected error: %v", prefix, err)
```
Comment (jrick) [inline waddrmgr/manager_test.go]: return or fatal

File: wallet/chainntfns.go [hunk 12]
```diff
@@ -148,6 +148,9 @@ func (w *Wallet) addReceivedTx(tx *btcutil.Tx, block *txstore.Block) error {
 			if err != nil {
 				return err
 			}
+			if err := w.markAddrsUsed(txr); err != nil {
```
Comment (jrick) [inline wallet/chainntfns.go]: This belongs in a different pr (or wait for intwtxmgr, which doesn't have this bug)

File: wallet/wallet.go [hunk 13]
```diff
@@ -720,7 +719,9 @@ func (w *Wallet) diskWriter() {
 // there are any transactions with outputs to this address in the blockchain or
 // the btcd mempool.
 func (w *Wallet) AddressUsed(addr waddrmgr.ManagedAddress) bool {
-	return addr.Used()
+	// TODO: handle err
```
Comment (jrick) [inline wallet/wallet.go]: Can you add this?

In fact, just remove this function.  The caller already has the managed address, so they can check this themselves.

File: waddrmgr/manager_test.go (hunk 10 above)
Comment (tuxcanfly) [inline waddrmgr/manager_test.go]: Added `return false` to all unexpected errs.

File: wallet/wallet.go (hunk 13 above)
Comment (tuxcanfly) [inline wallet/wallet.go]: OK, removed.

File: wallet/chainntfns.go (hunk 12 above)
Comment (tuxcanfly) [inline wallet/chainntfns.go]: OK, reverted the fix.

**Comment by jrick** [comment]
All of the waddrmgr changes look ok now but we need handling in Wallet.CurrentAddress so that if the error is because no addresses exist yet, it handles it by creating the first external address.

**Comment by tuxcanfly** [comment]
Updated `CurrentAddress` to handle `ErrAddressNotFound` and generate a new address.

**Comment by jrick** [comment]
Thanks, going to test this out but the diff looks ok now.  Can you squash?

**Comment by tuxcanfly** [comment]
Sure, done.

**Comment by jrick** [comment]
rebase on master too

**Comment by tuxcanfly** [comment]
OK, rebased.
